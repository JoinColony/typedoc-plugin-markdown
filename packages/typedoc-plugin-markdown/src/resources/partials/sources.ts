import { DeclarationReflection, SignatureReflection } from 'typedoc';

import { MarkdownThemeContext } from '../..';

import { heading, link } from '../../utils/elements';

export const sourcesPartial = (
  context: MarkdownThemeContext,
  model: DeclarationReflection | SignatureReflection,
) => {
  const md: string[] = [];

  if (model.implementationOf) {
    md.push(heading(4, 'Implementation of'));
    md.push(context.typeAndParentPartial(model.implementationOf));
  }

  if (model.inheritedFrom) {
    md.push(heading(4, 'Inherited from'));
    md.push(context.typeAndParentPartial(model.inheritedFrom));
  }

  if (model.overwrites) {
    md.push(heading(4, 'Overrides'));
    md.push(context.typeAndParentPartial(model.overwrites));
  }

  if (model.sources) {
    md.push(heading(4, 'Defined in'));
    model.sources.forEach((source) => {
      if (source.url) {
        md.push(link(`${source.fileName}:${source.line}`, source.url));
      } else {
        md.push(`${source.fileName}:${source.line}`);
      }
    });
  }
  return md.join('\n\n');
};
