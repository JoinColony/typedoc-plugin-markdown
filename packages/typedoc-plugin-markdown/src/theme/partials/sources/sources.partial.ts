import {
  ArrayType,
  DeclarationReflection,
  ReferenceType,
  SignatureReflection,
} from 'typedoc';
import { MarkdownThemeRenderContext } from '../../theme.context';
import { escapeChars, heading, link } from '../../theme.utils';

export const sourcesPartial = (
  context: MarkdownThemeRenderContext,
  model: DeclarationReflection | SignatureReflection,
) => {
  const md: string[] = [];

  if (model.implementationOf) {
    md.push(heading(4, 'Implementation of'));
    md.push(typeAndParent(context, model.implementationOf));
  }

  if (model.inheritedFrom) {
    md.push(heading(4, 'Inherited from'));
    md.push(typeAndParent(context, model.inheritedFrom));
  }

  if (model.overwrites) {
    md.push(heading(4, 'Overrides'));
    md.push(typeAndParent(context, model.overwrites));
  }

  if (model.sources) {
    md.push(heading(4, 'Defined in'));
    model.sources.forEach((source) => {
      if (source.url) {
        md.push(
          link(`${escapeChars(source.fileName)}:${source.line}`, source.url),
        );
      } else {
        md.push(`${escapeChars(source.fileName)}:${source.line}`);
      }
    });
  }
  return md.join('\n\n');
};

const typeAndParent = (
  context: MarkdownThemeRenderContext,
  props: ArrayType | ReferenceType,
) => {
  const getUrl = (name: string, url: string) =>
    `[${name}](${context.relativeURL(url)})`;
  if (props) {
    if ('elementType' in props) {
      return typeAndParent(context, props.elementType as any) + '[]';
    } else {
      if (props.reflection) {
        const md: string[] = [];
        if (props.reflection instanceof SignatureReflection) {
          if (props.reflection.parent?.parent?.url) {
            md.push(
              getUrl(
                props.reflection.parent.parent.name,
                props.reflection.parent.parent.url,
              ),
            );
            if (props.reflection.parent.url) {
              md.push(
                getUrl(
                  props.reflection.parent.name,
                  props.reflection.parent.url,
                ),
              );
            }
          }
        } else {
          if (props.reflection.parent?.url) {
            md.push(
              getUrl(props.reflection.parent.name, props.reflection.parent.url),
            );
            if (props.reflection.url) {
              md.push(getUrl(props.reflection.name, props.reflection.url));
            }
          }
        }
        return md.length > 0 ? md.join('.') : props.name;
      } else {
        return escapeChars(props.toString());
      }
    }
  }
  return 'void';
};
