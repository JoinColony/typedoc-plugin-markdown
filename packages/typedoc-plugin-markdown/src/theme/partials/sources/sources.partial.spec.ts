import { DeclarationReflection, ProjectReflection } from 'typedoc';
import { sourcesPartial } from './sources.partial';

const getProp = (reflection: DeclarationReflection) => {
  const prop = reflection.findReflectionByName('prop');
  prop.sources = prop.sources.map((source) => {
    delete source.url;
    return source;
  });
  return prop;
};

describe(`Sources Partial:`, () => {
  let project: ProjectReflection;

  beforeAll(async () => {
    project = global.getProject('sources.ts');
  });

  test(`should display implementation of sources'`, () => {
    expect(
      sourcesPartial(
        global.getMockContext(),
        getProp(
          project.findReflectionByName('SomeClass') as DeclarationReflection,
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should display inherited sources'`, () => {
    expect(
      sourcesPartial(
        global.getMockContext(),
        getProp(
          project.findReflectionByName(
            'AnotherInterface',
          ) as DeclarationReflection,
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should display overide sources'`, () => {
    expect(
      sourcesPartial(
        global.getMockContext(),
        getProp(
          project.findReflectionByName('AnotherClass') as DeclarationReflection,
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });
});
