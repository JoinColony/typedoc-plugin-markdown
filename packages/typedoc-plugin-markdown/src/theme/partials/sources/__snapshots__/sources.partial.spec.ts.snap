// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sources Partial: should display implementation of sources' 1`] = `
"#### Implementation of

SomeInterface.prop

#### Defined in

sources.ts:6"
`;

exports[`Sources Partial: should display inherited sources' 1`] = `
"#### Inherited from

SomeClass.prop

#### Defined in

sources.ts:2"
`;

exports[`Sources Partial: should display overide sources' 1`] = `
"#### Overrides

SomeClass.prop

#### Defined in

sources.ts:10"
`;
